app/workers/certificates.php

```diff
@@ -26,169 +28,224 @@ public function init(): void
+            if (!empty(App::getEnv('_APP_DOMAIN', ''))) {
+                $mainDomain = App::getEnv('_APP_DOMAIN', '');
+            } else {
+                $domainDocument = $dbForConsole->findOne('domains', [], 0, ['_id'], ['ASC']);
+                $mainDomain = $domainDocument ? $domainDocument->getAttribute('domain') : $domain->get();
             }
-        }
-
-        if ($validateCNAME) {
-            $validator = new CNAME($target->get()); // Verify Domain with DNS records
-
-            if(!$validator->isValid($domain->get())) {
-                throw new Exception('Failed to verify domain DNS records');
+    
+            // If not main domain, we will check CNAME record
+            $validateCNAME = false;
+            if ($domain->get() !== $mainDomain) {
+                $validateCNAME = true;
             }
-        }
-
-        $certificate = $dbForConsole->findOne('certificates', [
-            new Query('domain', QUERY::TYPE_EQUAL, [$domain->get()])
-        ]);
-
-        // $condition = ($certificate
-        //     && $certificate instanceof Document
-        //     && isset($certificate['issueDate'])
-        //     && (($certificate['issueDate'] + ($expiry)) > time())) ? 'true' : 'false';
-
-        // throw new Exception('cert issued at'.date('d.m.Y H:i', $certificate['issueDate']).' | renew date is: '.date('d.m.Y H:i', ($certificate['issueDate'] + ($expiry))).' | condition is '.$condition);
-
-        $certificate = (!empty($certificate) && $certificate instanceof $certificate) ? $certificate->getArrayCopy() : [];
-
-        if (
-            !empty($certificate)
-            && isset($certificate['issueDate'])
-            && (($certificate['issueDate'] + ($expiry)) > \time())
-        ) { // Check last issue time
-            throw new Exception('Renew isn\'t required');
-        }
-
-        $staging = (App::isProduction()) ? '' : ' --dry-run';
-        $email = App::getEnv('_APP_SYSTEM_SECURITY_EMAIL_ADDRESS');
-
-        if (empty($email)) {
-            throw new Exception('You must set a valid security email address (_APP_SYSTEM_SECURITY_EMAIL_ADDRESS) to issue an SSL certificate');
-        }
-
-        $stdout = '';
-        $stderr = '';
-
-        $exit = Console::execute("certbot certonly --webroot --noninteractive --agree-tos{$staging}"
-            . " --email " . $email
-            . " -w " . APP_STORAGE_CERTIFICATES
-            . " -d {$domain->get()}", '', $stdout, $stderr);
-
-        if ($exit !== 0) {
-            throw new Exception('Failed to issue a certificate with message: ' . $stderr);
-        }
-
-        $path = APP_STORAGE_CERTIFICATES . '/' . $domain->get();
-
-        if (!\is_readable($path)) {
-            if (!\mkdir($path, 0755, true)) {
-                throw new Exception('Failed to create path...');
+    
+            if (empty($domain->get())) {
+                throw new ExceptionCertificate('Missing certificate domain.');
```

Comment:
We should use a regular exception, and not trust the exception to handle the count for us. Resque is restarting the PHP script every execution so you can't trust the script memory.

Even with the new library and multiple threads the count might get mixed between different domains.

You should rely on the certificate document for counting the attempts and reseting them on every successful generation.

Reply:
Count has been refactored.